File: Data Structures & Algorithms/validate-parentheses/submission-1.cpp
class Solution {
public:
    bool isValid(string s) {
        // iterate through string. when you see a left bracket push it onto a stack
        // when you see the right bracket, visit top of stack and pop it
        // if the top isn't the corresponding left bracket return false
        // return false if you try to pop an empty stack

        stack<char> charstack;
        for (int i =0; i < s.size(); i++)
        {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[')
                charstack.push(s[i]);
            else if (charstack.size() == 0) return false;
            else
            {
                if (s[i] == ')')
                    {
                        if (charstack.top() != '(') return false;
                        else charstack.pop(); 
                    }
                if (s[i] == ']')
                    {
                        if (charstack.top() != '[') return false;
                        else charstack.pop(); 
                    }
                if (s[i] == '}')
                    {
                        if (charstack.top() != '{') return false;
                        else charstack.pop(); 
                    }
            }
            
        }

        if (charstack.size() != 0) return false;

        return true;
    }
};

File: Data Structures & Algorithms/valid-palindrome-ii/submission-1.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1
    
        if self.isValid(s):
            return True
        while l < r: 
            if s[l] != s[r]: 
                if self.isValid(s[:l] + s[l+1:]) or self.isValid(s[:r] + s[r+1:]):
                    return True
            
            l += 1
            r -= 1
        
        return False

    def isValid(self, s: str) -> bool:  
        l = 0
        r = len(s) - 1

        while l < r:
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        
        return True